include/ginkgo/core/base/array.hpp

```diff
@@ -65,6 +65,86 @@ void convert_data(std::shared_ptr<const Executor> exec, size_type size,
                   const SourceType* src, TargetType* dst);
 
 
+/**
+ * @internal
+ *
+ * Array-like non-owning wrapper for const data, to be used in conjunction with
+ * `array_const_cast` and `create_const` to create matrix type wrappers from
+ * constant data.
+ *
+ * @tparam ValueType  the type of elements stored in the array view.
+ */
+template <typename ValueType>
+class ConstArrayView {
+public:
+    /**
+     * The type of elements stored in the array view.
+     */
+    using value_type = ValueType;
+
+    /**
+     * Constructs an array view from existing data.
+     *
+     * @param exec  the executor in whose memory space the data resides.
+     * @param num_elems  the number of elements in this array view.
+     * @param data  a pointer to the first element of this array view.
+     */
+    ConstArrayView(std::shared_ptr<const Executor> exec, size_type num_elems,
+                   const ValueType* data)
+        : exec_{std::move(exec)}, num_elems_{num_elems}, data_{data}
+    {}
+
+    /*
+     * To avoid any collisions with the value semantics of normal arrays,
+     * disable assignment altogether.
+     */
+    ConstArrayView& operator=(const ConstArrayView&) = delete;
+    ConstArrayView& operator=(ConstArrayView&&) = delete;
```

Comment:
We could still allow move-assignment and nullify the internal pointer of the passed argument.